### ISSUE TITLE
NaN gradients from pfaffian computation when determinant is zero (x^0.25 singularity)

### ISSUE THREAD

**Issue by MarcoArmenta** [MatchCake/MatchCake#125]
## Summary

 When computing transition amplitudes via `|det(O)|^{1/4}` (fourth root of absolute determinant), zero or near-zero determinants produce **NaN gradients** that corrupt the entire backward pass. This affects any use case
  that takes a fractional power of the pfaffian/determinant for gradient-based optimization.

 ## Background

 In our downstream project [FermiDL](https://github.com/Mamarco-Armenta/FermiDL), we build weight matrices from non-interacting fermionic circuits using MatchCake. The amplitude `|<l|U(θ)|j>|` is computed as
 `|det(O)|^{1/4}` (since `Pf(O) = sqrt(|det(O)|)` and amplitude = `sqrt(Pf(O))`).

 Parity-forbidden transitions (where `popcount(l) ≠ popcount(j) mod 2`) produce determinants that are numerically zero or near-zero (~1e-32 residuals).

 ## The Problem

 The derivative of `f(x) = x^{0.25}` is:

 f'(x) = 0.25 * x^{-0.75} = 0.25 / x^{0.75}

 At `x = 0`, this is **infinite**. In PyTorch:

 1. Forward pass: `0^{0.25} = 0` ✓ (correct)
 2. Backward pass: `0.25 * 0^{-0.75}` = **NaN** ✗

 Even when these entries are later multiplied by zero (e.g., via a parity mask), the NaN propagates because `NaN * 0 = NaN` in IEEE 754 arithmetic. This corrupts all parameter gradients.

 ## Current State in MatchCake

 MatchCake's pfaffian functions use an additive epsilon approach:

 **`matchcake/utils/_pfaffian.py`** (lines ~300, 305, 333):
 ```python
 pf = sqrt(|det| + epsilon)  # epsilon=1e-12
 matchcake/utils/torch_pfaffian.py (line ~11):
 pf = torch.sqrt(torch.abs(det) + Pfaffian.EPSILON)  # EPSILON=1e-10
```
 This prevents NaN for sqrt (exponent 0.5), but does not protect downstream consumers who may take further fractional powers of the pfaffian (e.g., pf^{0.5} to get amplitudes, which is equivalent to |det|^{0.25}).

 Our Workaround.

 In FermiDL, we changed:

 # BEFORE (NaN gradients):
```python
amplitudes = torch.pow(torch.clamp(torch.abs(dets), min=0.0), 0.25)
```

 # AFTER (fixed):
```python
 amplitudes = torch.pow(torch.clamp(torch.abs(dets), min=1e-32), 0.25)
```

 The clamp floor of 1e-32 ensures gradients are large but finite (~10^{24}), and when multiplied by a zero parity mask, they cleanly produce 0.

 Suggested Fix

 Consider one or more of:

 1. Document the singularity: Warn users that fractional powers of pfaffians/determinants require an epsilon floor to avoid NaN gradients at zero.
 2. Add a clamp_min parameter to pfaffian functions (default 1e-32 or similar) so consumers don't have to handle this themselves.
 3. Provide a numerically safe amplitude function that computes |det|^{1/4} (or arbitrary |det|^{p}) with built-in gradient safety.
 4. Use torch.clamp instead of additive epsilon in existing pfaffian functions — sqrt(clamp(|det|, min=eps)) is mathematically cleaner than sqrt(|det| + eps) because it doesn't bias non-zero values.

 Environment

 - MatchCake version: 0.2.2 (dev branch, commit 1558e55)
 - PyTorch: >=2.6
 - Python: 3.12

**Comment by JeremieGince** [comment]
Thanks for opening this issue! I'll follow up with a PR soon.

Cheers

**Comment by JeremieGince** [comment]
Hi @MarcoArmenta ,

I've opened a PR for this issue: https://github.com/MatchCake/MatchCake/pull/128. Could you take a look and confirm it resolves the problem? I'm not entirely sure I've covered the exact edge case you're referring to.

Cheers
